docs/MIDDLEWARE_SPEC.md

```diff
@@ -0,0 +1,113 @@
+# Middelware Specification (v0.1.0)
+
+## Context
+
+The Singer specification is pretty mature and allows a lot of 
+end-users to extract and load data from one source to a 
+specific target very simply, and allows developers to easily
+implement new taps without worrying about targets and vice-versa.
+
+However, it is not straightforward, as of now, to run Singer channels (taps + targets)
+in a production environment.
+Because logs and metrics, by default when using the [singer-python](https://github.com/singer-io/singer-python) lib, 
+are send to [stderr](https://github.com/singer-io/singer-python/blob/master/singer/logging.conf#L12), it is very difficult
+and not straightforward to get to them and exploit them.
+
+To bring the specification one step further, we need to improve on this point.
+One possible implementation to resolve this problem, would be the implementation of a *middleware* type and *LOG* messages into the specification.
+
+## Definition
+
+### LOG Message
+
+A LOG message contain useful information relative to the run of the tap/broker/target.
+
+
+A LOG message have the following properties:
+
+- `type` **Required**. Always `LOG`
+- `level` **Required**. The level of the LOG message. Should be one of `CRITICAL`, `ERROR`, `WARNING`, `INFO`, `DEBUG`
+- `value` **Required**. The JSON formatted log
+- `tags` **Optionnal**. An object of tags
+
+The semantics of a LOG value are not part of the specification,
+and should be determined independently by each tap/broker/target.
+
+
+Example
+```
+{"type": "LOG", "level": "DEBUG", "value": {"message": "All is good there"}, "tags": {"endpoint": "users", "elastic": "true"}}
+{"type": "LOG", "level": "INFO", "value": {"type": "counter", "metric": "record_count", "value": 10000}, "tags": {"kind": "metric", "statsd": "true"}}
+{"type": "LOG", "level": "INFO", "value": {"type": "timer", "metric": "job_duration", "value": 10.0}, "tags": {"kind": "metric"}}
+{"type": "LOG", "level": "ERROR", "value": {"message": "I'm a teapot", "code": "418"}, "tags": {"pagerduty": "alert_1"}}
+```
+
+### Middleware
```

Comment:
This is a good suggestion. Its something we've been thinking about here as well.